setup.py

```diff
@@ -1,26 +1,164 @@
-import contextlib
 import io
 import os
 import re
 import subprocess
 from typing import List, Set
-import warnings
-from pathlib import Path
+import sys
+from typing import List
 
 from packaging.version import parse, Version
-import setuptools
+from setuptools import setup, find_packages, Extension
+from setuptools.command.build_ext import build_ext
+from shutil import which
 import torch
-import torch.utils.cpp_extension as torch_cpp_ext
-from torch.utils.cpp_extension import (BuildExtension, CUDAExtension,
-                                       CUDA_HOME, ROCM_HOME)
+from torch.utils.cpp_extension import CUDA_HOME
 
 ROOT_DIR = os.path.dirname(__file__)
 
 MAIN_CUDA_VERSION = "12.1"
 
-# Supported NVIDIA GPU architectures.
-NVIDIA_SUPPORTED_ARCHS = {"6.1", "7.0", "7.5", "8.0", "8.6", "8.9", "9.0"}
-ROCM_SUPPORTED_ARCHS = {"gfx90a", "gfx942", "gfx1100"}
+def is_sccache_available() -> bool:
+    return which("sccache") is not None
+
+
+def is_ccache_available() -> bool:
+    return which("ccache") is not None
+
+def is_ninja_available() -> bool:
+    return which("ninja") is not None
+
+
+def remove_prefix(text, prefix):
+    if text.startswith(prefix):
+        return text[len(prefix):]
+    return text
+
+
+class CMakeExtension(Extension):
+
+    def __init__(self, name: str, cmake_lists_dir: str = '.', **kwa) -> None:
+        super().__init__(name, sources=[], **kwa)
+        self.cmake_lists_dir = os.path.abspath(cmake_lists_dir)
+
+
+class cmake_build_ext(build_ext):
+    # A dict of extension directories that have been configured.
+    did_config = {}
+
+    # Determine number of compilation jobs and optionally nvcc compile threads.
+    def compute_num_jobs(self):
+        try:
+            # os.sched_getaffinity() isn't universally available, so fall back
+            # to os.cpu_count() if we get an error here.
+            num_jobs = len(os.sched_getaffinity(0))
+        except AttributeError:
+            num_jobs = os.cpu_count()
+
+        nvcc_cuda_version = get_nvcc_cuda_version()
+        if nvcc_cuda_version >= Version("11.2"):
+            nvcc_threads = int(os.getenv("NVCC_THREADS", 8))
+            num_jobs = max(1, round(num_jobs / (nvcc_threads / 4)))
+        else:
+            nvcc_threads = None
+
+        return num_jobs, nvcc_threads
+
+    # Perform cmake configuration for a single extension.
+    def configure(self, ext: CMakeExtension) -> None:
```

Comment:
This only gets run once due to the early return being hit on every subsequent extension. Therefor I suggest removing the `ext` parameter and then removing the `did_config` checks, changing outdir to `os.path.abspath(os.path.join(self.build_lib, "aphrodite"))` so that it stays the same